packages/aws-cdk/lib/api/toolkit-info.ts

```diff
@@ -1,44 +1,126 @@
       debug('The environment %s doesn\'t have the CDK toolkit stack (%s) installed. Use %s to setup your environment for use with the toolkit.',
         environment.name, stackName, colors.blue(`cdk bootstrap "${environment.name}"`));
-      return undefined;
+      return ToolkitInfo.bootstrapStackNotFoundInfo(sdk);
     }
     if (stack.stackStatus.isCreationFailure) {
       // Treat a "failed to create" bootstrap stack as an absent one.
       debug('The environment %s has a CDK toolkit stack (%s) that failed to create. Use %s to try provisioning it again.',
         environment.name, stackName, colors.blue(`cdk bootstrap "${environment.name}"`));
-      return undefined;
+      return ToolkitInfo.bootstrapStackNotFoundInfo(sdk);
     }
 
-    return new ToolkitInfo(stack, sdk);
+    return new ExistingToolkitInfo(stack, sdk);
+  }
+
+  public static fromStack(stack: CloudFormationStack, sdk: ISDK): ToolkitInfo {
+    return new ExistingToolkitInfo(stack, sdk);
+  }
+
+  public static bootstraplessDeploymentsOnly(sdk: ISDK): ToolkitInfo {
+    return new BootstrapStackNotFoundInfo(sdk, 'Trying to perform an operation that requires a bootstrap stack; you should not see this error, this is a bug in the CDK CLI.');
+  }
+
+  public static bootstrapStackNotFoundInfo(sdk: ISDK): ToolkitInfo {
+    return new BootstrapStackNotFoundInfo(sdk, 'This deployment requires a bootstrap stack with a known name; pass \'--toolkit-stack-name\' or switch to using the \'DefaultStackSynthesizer\' (see https://docs.aws.amazon.com/cdk/latest/guide/bootstrapping.html)');
   }
 
-  constructor(public readonly stack: CloudFormationStack, private readonly sdk?: ISDK) {
+  public abstract readonly found: boolean;
+  public abstract readonly bucketUrl: string;
+  public abstract readonly bucketName: string;
+  public abstract readonly version: number;
+  public abstract readonly bootstrapStack: CloudFormationStack;
+
+  private readonly ssmCache = new Map<string, number>();
+
+  constructor(protected readonly sdk: ISDK) {
+  }
+  public abstract validateVersion(expectedVersion: number, ssmParameterName: string | undefined): Promise<void>;
+  public abstract prepareEcrRepository(repositoryName: string): Promise<EcrRepositoryInfo>;
+
+  /**
+   * Read a version from an SSM parameter, cached
+   */
+  protected async versionFromSsmParameter(parameterName: string): Promise<number> {
+    const existing = this.ssmCache.get(parameterName);
+    if (existing !== undefined) { return existing; }
+
+    const ssm = this.sdk.ssm();
+
+    try {
+      const result = await ssm.getParameter({ Name: parameterName }).promise();
+
+      const asNumber = parseInt(`${result.Parameter?.Value}`, 10);
+      if (isNaN(asNumber)) {
+        throw new Error(`SSM parameter ${parameterName} not a number: ${result.Parameter?.Value}`);
+      }
+
+      this.ssmCache.set(parameterName, asNumber);
+      return asNumber;
+    } catch (e) {
+      if (e.code === 'ParameterNotFound') {
```

Comment:
Are we sure this only happens when it's a legitimate 404/NotFound? I would bet in some circumstances (e.g., cross-account) at least, you get a ParameterNotFound when you don't have access. That would mean the user gets the less helpful version of the message.

Reply:
After thinking about this a little: we'd always assume the cross-account role first, then do an in-account read of this parameter. So combined with the research you did, seems like we're safe.